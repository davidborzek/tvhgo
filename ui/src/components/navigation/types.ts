import { ReactElement } from 'react';

export type INavigationItem = {
  to: string;
  title: string;
  requiredRoles?: string[];
  icon?: ReactElement;
  items?: INavigationItem[];
};
