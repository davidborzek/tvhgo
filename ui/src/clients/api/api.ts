import axios from 'axios';
import {
  Channel,
  CreateTokenResponse,
  EpgChannel,
  EpgEvent,
  ErrorResponse,
  ListResponse,
  Recording,
  Session,
  Token,
  TwoFactorAuthSettings,
  TwoFactorAuthSetupResult,
  UpdateRecording,
  UpdateUser,
  UpdateUserPassword,
  UserResponse,
} from './api.types';
import qs from 'qs';

type PaginationQuery = {
  limit?: number;
  offset?: number;
};

type SortQuery = {
  sort_key?: string;
  sort_dir?: string;
};

type PaginationSortQuery = PaginationQuery & SortQuery;

export type GetEpgEventsQuery = PaginationSortQuery & {
  title?: string;
  lang?: string;
  channel?: string;
  contentType?: string;
  fullText?: boolean;
  nowPlaying?: boolean;
  durationMin?: number;
  durationMax?: number;
  startsAt?: number;
  endsAt?: number;
};

export type GetEpgChannelEventsQuery = PaginationSortQuery & {
  startsAt?: number;
  endsAt?: number;
};

export type RecordingStatus = 'upcoming' | 'finished' | 'failed' | 'removed';

export type GetRecordingsQuery = PaginationSortQuery & {
  status?: RecordingStatus;
};

export type GetChannelsQuery = PaginationSortQuery & {
  name?: string;
};

export class ApiError extends Error {
  constructor(
    public readonly code: number,
    message: string
  ) {
    super(message);
  }
}

const client = axios.create({
  baseURL: '/api',
  paramsSerializer: (params) => qs.stringify(params, { arrayFormat: 'repeat' }),
});

client.interceptors.response.use(
  (res) => res,
  (err) => {
    if (axios.isAxiosError(err) && err.response) {
      const data = err.response.data as ErrorResponse;
      return Promise.reject(new ApiError(err.response.status, data.message));
    }
    return Promise.reject(err);
  }
);

export async function login(username: string, password: string, code?: string) {
  await client.post('/login', {
    username,
    password,
    totp: code,
  });
}

export async function logout(): Promise<void> {
  await client.post('/logout');
}

export async function getUser(): Promise<UserResponse> {
  const response = await client.get<UserResponse>('/user');
  return response.data;
}

export async function getEpgEvents(
  q?: GetEpgEventsQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>('/epg/events', {
    params: q,
  });
  return response.data;
}

export async function getRelatedEpgEvents(
  id: number,
  q?: PaginationSortQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>(
    `/epg/events/${id}/related`,
    {
      params: q,
    }
  );
  return response.data;
}

export async function getEpgEvent(id: number): Promise<EpgEvent> {
  const response = await client.get<EpgEvent>(`/epg/events/${id}`);
  return response.data;
}

export async function getEpg(
  q?: GetEpgChannelEventsQuery
): Promise<Array<EpgChannel>> {
  const response = await client.get<Array<EpgChannel>>('/epg', {
    params: q,
  });
  return response.data;
}

export async function getChannel(id: string): Promise<Channel> {
  const response = await client.get<Channel>(`/channels/${id}`);
  return response.data;
}

export async function getChannels(
  q?: GetChannelsQuery
): Promise<Array<Channel>> {
  const response = await client.get<Array<Channel>>(`/channels`, {
    params: q,
  });
  return response.data;
}

export async function fetch<T>(path: string): Promise<T> {
  const response = await client.get<T>(path);
  return response.data;
}

export async function recordByEvent(
  eventId: number,
  configId?: string
): Promise<void> {
  await client.post('/recordings/event', {
    eventId,
    configId,
  });
}

export async function stopRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/stop`);
}

export async function cancelRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/cancel`);
}

export async function removeRecording(id: string): Promise<void> {
  await client.delete(`/recordings/${id}`);
}

export async function stopRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/stop`, null, {
    params: { ids },
  });
}

export async function cancelRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/cancel`, null, {
    params: { ids },
  });
}

export async function removeRecordings(ids: string[]): Promise<void> {
  await client.delete(`/recordings`, {
    params: { ids },
  });
}

export async function getRecordings(
  q?: GetRecordingsQuery
): Promise<ListResponse<Recording>> {
  const response = await client.get<ListResponse<Recording>>(`/recordings`, {
    params: q,
  });
  return response.data;
}

export async function getRecording(id: string): Promise<Recording> {
  const response = await client.get<Recording>(`/recordings/${id}`);
  return response.data;
}

export async function updateRecording(
  id: string,
  opts: UpdateRecording
): Promise<void> {
  await client.patch(`/recordings/${id}`, opts);
}

export async function updateUser(opts: UpdateUser): Promise<UserResponse> {
  const response = await client.patch<UserResponse>(`/user`, opts);
  return response.data;
}

export async function updateUserPassword(
  opts: UpdateUserPassword
): Promise<void> {
  await client.patch(`/user/password`, opts);
}

export function getRecordingUrl(id: string): string {
  return `/api/recordings/${id}/stream`;
}

export async function getSessions(): Promise<Array<Session>> {
  const response = await client.get<Array<Session>>(`/sessions`);
  return response.data;
}

export async function deleteSession(id: number): Promise<void> {
  return await client.delete(`/sessions/${id}`);
}

export async function getTwoFactorAuthSettings(): Promise<TwoFactorAuthSettings> {
  const response = await client.get<TwoFactorAuthSettings>(`/two-factor-auth`);
  return response.data;
}

export async function deactivateTwoFactorAuth(code: string): Promise<void> {
  await client.put(`/two-factor-auth/deactivate`, {
    code,
  });
}

export async function setupTwoFactorAuth(
  password: string
): Promise<TwoFactorAuthSetupResult> {
  const response = await client.put<TwoFactorAuthSetupResult>(
    `/two-factor-auth/setup`,
    {
      password,
    }
  );
  return response.data;
}

export async function activateTwoFactorAuth(
  password: string,
  code: string
): Promise<void> {
  await client.put(`/two-factor-auth/activate`, {
    password,
    code,
  });
}

export async function getTokens(): Promise<Array<Token>> {
  const response = await client.get<Array<Token>>(`/tokens`);
  return response.data;
}

export async function deleteToken(id: number): Promise<void> {
  await client.delete(`/tokens/${id}`);
}

export async function createToken(name: string): Promise<CreateTokenResponse> {
  const response = await client.post<CreateTokenResponse>(`/tokens`, { name });
  return response.data;
}
