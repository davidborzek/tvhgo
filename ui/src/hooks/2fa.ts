import { useEffect, useState } from 'react';
import { TwoFactorAuthSettings } from '@/clients/api/api.types';
import { useLoading } from '@/contexts/LoadingContext';
import {
  ApiError,
  activateTwoFactorAuth,
  deactivateTwoFactorAuth,
  getTwoFactorAuthSettings,
  setupTwoFactorAuth,
} from '@/clients/api/api';
import { useTranslation } from 'react-i18next';
import { useNotification } from './notification';

export const useTwoFactorAuthSettings = () => {
  const [twoFactorAuthSettings, setTwoFactorAuthSettings] =
    useState<TwoFactorAuthSettings | null>(null);

  const { setIsLoading } = useLoading();

  const fetchTwoFactorAuthSettings = async () => {
    setIsLoading(true);
    return await getTwoFactorAuthSettings()
      .then(setTwoFactorAuthSettings)
      .finally(() => setIsLoading(false));
  };

  useEffect(() => {
    fetchTwoFactorAuthSettings();
  }, []);

  return {
    fetchTwoFactorAuthSettings,
    twoFactorAuthSettings,
  };
};

export const useSetupTwoFactorAuth = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('setupTwoFactorAuth');

  const { t } = useTranslation();

  const [loading, setLoading] = useState<boolean>(false);
  const [twoFactorUrl, setTwoFactorUrl] = useState<string | null>(null);

  return {
    setupTwoFactorAuth: async (password: string) => {
      dismissNotification();
      setLoading(true);
      return await setupTwoFactorAuth(password)
        .then(({ url }) => setTwoFactorUrl(url))
        .catch((error) => {
          if (
            error instanceof ApiError &&
            error.code === 400 &&
            error.message === 'confirmation password is invalid'
          ) {
            notifyError(t('invalid_current_password'));
          } else if (error.message === 'two factor auth is already enabled') {
            notifyError(t('two_factor_auth_already_enabled'));
          } else {
            notifyError(t('unexpected'));
          }

          throw error;
        })
        .finally(() => setLoading(false));
    },
    activateTwoFactorAuth: async (password: string, code: string) => {
      dismissNotification();
      setLoading(true);

      return await activateTwoFactorAuth(password, code)
        .then(() => {
          notifySuccess(t('two_factor_auth_successfully_enabled'));
        })
        .catch((error) => {
          if (error instanceof ApiError && error.code === 400) {
            if (error.message === 'confirmation password is invalid') {
              notifyError(t('invalid_current_password'));
            } else if (error.message === 'invalid two factor code provided') {
              notifyError(t('invalid_verification_code'));
            } else {
              notifyError(t('unexpected'));
            }
          } else {
            notifyError(t('unexpected'));
          }

          throw error;
        })
        .finally(() => setLoading(false));
    },
    setTwoFactorUrl,
    twoFactorUrl,
    loading,
  };
};

export const useDeactivateTwoFactorAuth = () => {
  const { notifyError, notifySuccess, dismissNotification } = useNotification(
    'deactivateTwoFactorAuth'
  );

  const { t } = useTranslation();

  const [loading, setLoading] = useState<boolean>(false);

  return {
    deactivateTwoFactorAuth: async (code: string) => {
      dismissNotification();
      setLoading(true);

      return await deactivateTwoFactorAuth(code)
        .then(() => notifySuccess(t('two_factor_auth_successfully_disabled')))
        .catch((error) => {
          if (
            error instanceof ApiError &&
            error.code === 400 &&
            error.message === 'invalid two factor code provided'
          ) {
            notifyError(t('invalid_verification_code'));
          } else {
            notifyError(t('unexpected'));
          }

          throw error;
        })
        .finally(() => setLoading(false));
    },
    loading,
  };
};
