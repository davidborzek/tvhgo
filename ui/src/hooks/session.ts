import { ApiError, deleteSession } from '@/clients/api/api';
import { useTranslation } from 'react-i18next';
import { useNotification } from './notification';
import { useRevalidator } from 'react-router-dom';

export const useManageSessions = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('manageSessions');
  const revalidator = useRevalidator();

  const { t } = useTranslation();

  const _revokeSession = async (id: number) => {
    dismissNotification();

    return await deleteSession(id)
      .then(() => {
        notifySuccess(t('session_revoked'));
        revalidator.revalidate();
      })
      .catch((error) => {
        if (
          error instanceof ApiError &&
          error.code === 400 &&
          error.message === 'current session cannot be revoked'
        ) {
          notifyError(t('current_session_cannot_be_revoked'));
        } else {
          notifyError(t('unexpected'));
        }
      });
  };

  return {
    revokeSession: _revokeSession,
  };
};
