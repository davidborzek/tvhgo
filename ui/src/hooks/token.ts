import { useLoading } from '@/contexts/LoadingContext';
import { useEffect, useState } from 'react';
import { createToken, deleteToken, getTokens } from '@/clients/api/api';
import { useTranslation } from 'react-i18next';
import { Token } from '@/clients/api/api.types';
import { useNotification } from './notification';

export const useManageTokens = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('manageTokens');

  const [error, setError] = useState<string | null>(null);
  const [tokens, setTokens] = useState<Array<Token>>([]);

  const { t } = useTranslation();

  const { setIsLoading } = useLoading();

  const _getTokens = async () => {
    setIsLoading(true);
    return await getTokens()
      .then((sessions) => {
        setTokens(sessions);
      })
      .catch(() => {
        setError(t('unexpected'));
      })
      .finally(() => setIsLoading(false));
  };

  const _revokeToken = async (id: number) => {
    dismissNotification();
    setIsLoading(true);

    return await deleteToken(id)
      .then(() => {
        _getTokens();
        notifySuccess(t('session_revoked'));
      })
      .catch((error) => {
        notifyError(t('unexpected'));
      })
      .finally(() => setIsLoading(false));
  };

  useEffect(() => {
    _getTokens();
  }, []);

  return {
    getTokens: _getTokens,
    revokeToken: _revokeToken,
    tokens,
    error,
  };
};

export const useCreateToken = () => {
  const { notifyError, dismissNotification } = useNotification('createToken');

  const [token, setToken] = useState('');

  const { t } = useTranslation();

  const { setIsLoading } = useLoading();

  const _createToken = async (name: string) => {
    dismissNotification();
    setIsLoading(true);

    return await createToken(name)
      .then((res) => {
        setToken(res.token);
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => setIsLoading(false));
  };

  return {
    createToken: _createToken,
    setToken,
    token,
  };
};
