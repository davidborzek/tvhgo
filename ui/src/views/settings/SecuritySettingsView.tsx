import { useFormik } from 'formik';
import { useEffect, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { useNavigate, useLocation, Outlet, useLoaderData, useRevalidator } from 'react-router-dom';
import * as Yup from 'yup';

import Button from '@/components/common/button/Button';
import Headline from '@/components/common/headline/Headline';
import Input from '@/components/common/input/Input';
import useFormikErrorFocus from '@/hooks/formik';
import { useManageSessions } from '@/hooks/session';
import { useManageTokens } from '@/hooks/token';
import { useUpdateUserPassword } from '@/hooks/user';
import Form from '@/components/common/form/Form';

import styles from './SettingsView.module.scss';
import SessionList from '@/components/settings/sessionList/SessionList';
import TokenList from '@/components/settings/tokenList/TokenList';
import TwoFactorAuthSettingsOverview from '@/components/settings/twoFactorAuthSettings/TwoFactorAuthSettingsOverview';
import {
  getSessions,
  getTokens,
  getTwoFactorAuthSettings,
} from '@/clients/api/api';
import { Token, TwoFactorAuthSettings } from '@/clients/api/api.types';
import { Session } from '@/clients/api/api.types';

export enum SecuritySettingsRefreshStates {
  TWOFA = 'refresh_2fa',
  TOKEN = 'refresh_token',
}

export async function loader() {
  return Promise.all([getTwoFactorAuthSettings(), getSessions(), getTokens()]);
}

export const Component = () => {
  const { t } = useTranslation();
  const revalidator = useRevalidator();
  const navigate = useNavigate();
  const { state } = useLocation();
  const { revokeSession } = useManageSessions();
  const { revokeToken } = useManageTokens();
  const { updatePassword } = useUpdateUserPassword();

  const [twoFactorAuthSettings, sessions, tokens] = useLoaderData() as [TwoFactorAuthSettings, Array<Session>, Array<Token>]

  const currentPasswordRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const passwordRepeatRef = useRef<HTMLInputElement>(null);

  const passwordChangeValidationSchema = Yup.object().shape({
    currentPassword: Yup.string().required(t('input_required')),
    password: Yup.string()
      .required(t('input_required'))
      .min(8, t('password_min_chars_error')),
    passwordRepeat: Yup.string()
      .required(t('passwords_do_not_match'))
      .oneOf([Yup.ref('password')], t('passwords_do_not_match')),
  });

  const passwordChangeFormik = useFormik({
    initialValues: {
      currentPassword: '',
      password: '',
      passwordRepeat: '',
    },
    validationSchema: passwordChangeValidationSchema,
    onSubmit: async ({ password, currentPassword }) => {
      updatePassword({ password, currentPassword }).then((success) => {
        // TODO: can this be done better
        currentPasswordRef.current?.blur();
        passwordRef.current?.blur();
        passwordRepeatRef.current?.blur();
        success && passwordChangeFormik.resetForm();
      });
    },
  });

  useFormikErrorFocus(
    passwordChangeFormik,
    currentPasswordRef,
    passwordRef,
    passwordRepeatRef
  );

  useEffect(() => {
    switch (state) {
      case SecuritySettingsRefreshStates.TWOFA:
        revalidator.revalidate();
        break;
      case SecuritySettingsRefreshStates.TOKEN:
        revalidator.revalidate();
        break;
    }
  }, [state]);

  return (
    <>
      <div className={styles.row}>
        <Form
          onSubmit={passwordChangeFormik.handleSubmit}
          className={styles.section}
        >
          <Headline>{t('password')}</Headline>
          <Input
            placeholder={t('current_password')}
            label={t('current_password')}
            value={passwordChangeFormik.values.currentPassword}
            name="currentPassword"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.currentPassword
                ? passwordChangeFormik.errors.currentPassword
                : undefined
            }
            ref={currentPasswordRef}
            fullWidth
          />
          <Input
            placeholder={t('password')}
            label={t('password')}
            value={passwordChangeFormik.values.password}
            name="password"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.password
                ? passwordChangeFormik.errors.password
                : undefined
            }
            ref={passwordRef}
            fullWidth
          />
          <Input
            label={t('password_repeat')}
            placeholder={t('password_repeat')}
            value={passwordChangeFormik.values.passwordRepeat}
            name="passwordRepeat"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.passwordRepeat
                ? passwordChangeFormik.errors.passwordRepeat
                : undefined
            }
            ref={passwordRepeatRef}
            fullWidth
          />
          <div>
            <Button type="submit" label={t('save')} />
          </div>
        </Form>
      </div>
      <div className={styles.row}>
        <TwoFactorAuthSettingsOverview
          settings={twoFactorAuthSettings}
          onDisable={() =>
            navigate('/settings/security/two-factor-auth/disable')
          }
          onEnable={() => navigate('/settings/security/two-factor-auth/setup')}
        />
      </div>
      <div className={styles.row}>
        <SessionList sessions={sessions} onRevoke={revokeSession} />
      </div>
      <div className={styles.row}>
        <TokenList tokens={tokens} onRevoke={revokeToken} />
      </div>
      <Outlet />
    </>
  );
};

Component.displayName = 'SecuritySettingsView';
