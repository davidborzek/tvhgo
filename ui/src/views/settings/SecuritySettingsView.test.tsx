import { Session, Token, TwoFactorAuthSettings } from '@/clients/api/api.types';
import { useTwoFactorAuthSettings } from '@/hooks/2fa';
import { useManageSessions } from '@/hooks/session';
import { useManageTokens } from '@/hooks/token';
import { useUpdateUserPassword } from '@/hooks/user';
import { cleanup, render } from '@testing-library/react';
import { useLocation, useNavigate } from 'react-router-dom';
import { afterEach, beforeEach, expect, test, vi, describe } from 'vitest';
import SecuritySettingsView from './SecuritySettingsView';
import { userEvent } from '@testing-library/user-event';

vi.mock('react-router-dom');

vi.mock('@/hooks/session');
vi.mock('@/hooks/2fa');
vi.mock('@/hooks/token');
vi.mock('@/hooks/user');

const sessions: Session[] = [
  {
    id: 1,
    clientIp: '10.0.0.1',
    userAgent:
      'Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0',
    userId: 1,
    createdAt: 0,
    lastUsedAt: 0,
  },
  {
    id: 2,
    clientIp: '10.0.0.2',
    userAgent:
      'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36',
    userId: 1,
    createdAt: 0,
    lastUsedAt: 0,
  },
];

const tokens: Token[] = [
  {
    id: 1,
    name: 'My Token 1',
    createdAt: 0,
    updatedAt: 0,
  },
  {
    id: 2,
    name: 'My Token 2',
    createdAt: 0,
    updatedAt: 0,
  },
];

const twoFactorAuthEnabled: TwoFactorAuthSettings = {
  enabled: true,
};

const twoFactorAuthDisabled: TwoFactorAuthSettings = {
  enabled: false,
};

afterEach(() => {
  vi.restoreAllMocks();
  cleanup();
});

const navigateMock = vi.fn();
const updatePasswordMock = vi.fn();
const revokeSessionMock = vi.fn();
const revokeTokenMock = vi.fn();

beforeEach(() => {
  vi.mocked(useLocation).mockReturnValue({
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: null,
  });

  vi.mocked(useNavigate).mockReturnValue(navigateMock);

  vi.mocked(useUpdateUserPassword).mockReturnValue({
    updatePassword: updatePasswordMock,
  });
  updatePasswordMock.mockResolvedValue(true);

  vi.mocked(useManageSessions).mockReturnValue({
    error: null,
    getSessions: vi.fn(),
    revokeSession: revokeSessionMock,
    sessions,
  });

  vi.mocked(useManageTokens).mockReturnValue({
    error: null,
    getTokens: vi.fn(),
    revokeToken: revokeTokenMock,
    tokens,
  });

  vi.mocked(useTwoFactorAuthSettings).mockReturnValue({
    fetchTwoFactorAuthSettings: vi.fn(),
    twoFactorAuthSettings: twoFactorAuthEnabled,
  });
});

test('should render', () => {
  const document = render(<SecuritySettingsView />);
  expect(document.asFragment()).toMatchSnapshot();
});

describe('update password', () => {
  test('should update password', async () => {
    const document = render(<SecuritySettingsView />);

    const currentPassword = '12345678';
    const password = '87654321';

    const currenPasswordInput = document.container.querySelector(
      'input[name=currentPassword]'
    ) as Element;

    const passwordInput = document.container.querySelector(
      'input[name=password]'
    ) as Element;

    const passwordRepeatInput = document.container.querySelector(
      'input[name=passwordRepeat]'
    ) as Element;

    const saveButton = document.getByText('save');

    await userEvent.type(currenPasswordInput, currentPassword);
    await userEvent.type(passwordInput, password);
    await userEvent.type(passwordRepeatInput, password);

    expect(currenPasswordInput).toHaveValue(currentPassword);
    expect(passwordInput).toHaveValue(password);
    expect(passwordRepeatInput).toHaveValue(password);

    await userEvent.click(saveButton);

    expect(updatePasswordMock).toBeCalledWith({ currentPassword, password });

    expect(currenPasswordInput).toHaveValue('');
    expect(passwordInput).toHaveValue('');
    expect(passwordRepeatInput).toHaveValue('');
  });

  test('should not update password for all empty inputs', async () => {
    const document = render(<SecuritySettingsView />);

    const saveButton = document.getByText('save');
    await userEvent.click(saveButton);

    expect(updatePasswordMock).not.toBeCalled();

    const inputRequiredError = document.getAllByText('input_required');
    expect(inputRequiredError).toHaveLength(2);

    const passwordDoNotMatch = document.getByText('passwords_do_not_match');
    expect(passwordDoNotMatch).toBeInTheDocument();
  });

  test('should not update password for some empty inputs', async () => {
    const document = render(<SecuritySettingsView />);

    const password = '87654321';

    const passwordInput = document.container.querySelector(
      'input[name=password]'
    ) as Element;

    const saveButton = document.getByText('save');

    await userEvent.type(passwordInput, password);
    expect(passwordInput).toHaveValue(password);

    await userEvent.click(saveButton);
    expect(passwordInput).toHaveValue(password);

    expect(updatePasswordMock).not.toBeCalled();

    const inputRequiredError = document.getByText('input_required');
    expect(inputRequiredError).toBeInTheDocument();

    const passwordDoNotMatch = document.getByText('passwords_do_not_match');
    expect(passwordDoNotMatch).toBeInTheDocument();
  });

  test('should not update password when passwords do not match', async () => {
    const document = render(<SecuritySettingsView />);

    const currentPassword = '12345678';
    const password = '87654321';

    const currenPasswordInput = document.container.querySelector(
      'input[name=currentPassword]'
    ) as Element;

    const passwordInput = document.container.querySelector(
      'input[name=password]'
    ) as Element;

    const passwordRepeatInput = document.container.querySelector(
      'input[name=passwordRepeat]'
    ) as Element;
    const saveButton = document.getByText('save');

    await userEvent.type(currenPasswordInput, currentPassword);
    await userEvent.type(passwordInput, password);
    await userEvent.type(passwordRepeatInput, 'invalid');

    await userEvent.click(saveButton);
    expect(passwordInput).toHaveValue(password);

    expect(updatePasswordMock).not.toBeCalled();

    const inputRequiredError = document.queryByText('input_required');
    expect(inputRequiredError).not.toBeInTheDocument();

    const passwordDoNotMatch = document.getByText('passwords_do_not_match');
    expect(passwordDoNotMatch).toBeInTheDocument();
  });

  test('should not update password when password is invalid', async () => {
    const document = render(<SecuritySettingsView />);

    const currentPassword = '12345678';
    const password = '1234';

    const currenPasswordInput = document.container.querySelector(
      'input[name=currentPassword]'
    ) as Element;

    const passwordInput = document.container.querySelector(
      'input[name=password]'
    ) as Element;

    const passwordRepeatInput = document.container.querySelector(
      'input[name=passwordRepeat]'
    ) as Element;
    const saveButton = document.getByText('save');

    await userEvent.type(currenPasswordInput, currentPassword);
    await userEvent.type(passwordInput, password);
    await userEvent.type(passwordRepeatInput, password);

    await userEvent.click(saveButton);
    expect(passwordInput).toHaveValue(password);

    expect(updatePasswordMock).not.toBeCalled();

    const inputRequiredError = document.queryByText('input_required');
    expect(inputRequiredError).not.toBeInTheDocument();

    const passwordDoNotMatch = document.queryByText('passwords_do_not_match');
    expect(passwordDoNotMatch).not.toBeInTheDocument();

    const passwordInvalid = document.queryByText('password_min_chars_error');
    expect(passwordInvalid).toBeInTheDocument();
  });
});

test.each(sessions)('should revoke session with id $id', async (session) => {
  const document = render(<SecuritySettingsView />);

  const revokeSessionButtons = document.getAllByTestId('revoke-session-btn');
  expect(revokeSessionButtons).toHaveLength(sessions.length);

  await userEvent.click(revokeSessionButtons[sessions.indexOf(session)]);

  expect(revokeSessionMock).toBeCalledWith(session.id);
});

test.each(tokens)('should revoke token with id $id', async (token) => {
  const document = render(<SecuritySettingsView />);

  const revokeTokenButtons = document.getAllByTestId('revoke-token-btn');
  expect(revokeTokenButtons).toHaveLength(tokens.length);

  await userEvent.click(revokeTokenButtons[tokens.indexOf(token)]);

  expect(revokeTokenMock).toBeCalledWith(token.id);
});
