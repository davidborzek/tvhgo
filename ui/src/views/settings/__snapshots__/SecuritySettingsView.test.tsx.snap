// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="_row_65046f"
  >
    <form
      class="_Form_27c918 _section_65046f"
    >
      <h2
        class="_headline_8da35a"
      >
        password
      </h2>
      <div
        class="_inputContainer_ae92e9 "
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="currentPassword"
        >
          current_password
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9   "
            name="currentPassword"
            placeholder="current_password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9 "
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="password"
        >
          password
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9   "
            name="password"
            placeholder="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9 "
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="passwordRepeat"
        >
          password_repeat
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9   "
            name="passwordRepeat"
            placeholder="password_repeat"
            type="password"
            value=""
          />
        </div>
      </div>
      <div>
        <button
          class="_button_a7f9df   "
          type="submit"
        >
          save
        </button>
      </div>
    </form>
  </div>
  <div
    class="_row_65046f"
  >
    <div
      class="_settings_2131bd"
    >
      <h2
        class="_headline_8da35a"
      >
        two_factor_auth
      </h2>
      <button
        class="_button_a7f9df _red_a7f9df  "
      >
        disable
      </button>
    </div>
  </div>
  <div
    class="_row_65046f"
  >
    <div
      class="_SessionList_542de3"
    >
      <h2
        class="_headline_8da35a"
      >
        sessions
      </h2>
      <div
        class="_tableContainer_542de3"
      >
        <table
          class="_table_542de3"
        >
          <thead>
            <tr>
              <th>
                last_seen
              </th>
              <th
                class="_created_542de3"
              >
                created
              </th>
              <th
                class="_ip_542de3"
              >
                ip_address
              </th>
              <th>
                browser_and_os
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                date
              </td>
              <td
                class="_created_542de3"
              >
                date
              </td>
              <td
                class="_ip_542de3"
              >
                10.0.0.1
              </td>
              <td>
                user_agent_full
              </td>
              <td>
                <button
                  class="_button_a7f9df _red_a7f9df  "
                  data-testid="revoke-session-btn"
                >
                  <svg
                    class="_sessionDeleteButton_542de3"
                    fill="none"
                    height="800px"
                    viewBox="0 0 24 24"
                    width="800px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      id="Menu / Close_LG"
                    >
                      <path
                        d="M21 21L12 12M12 12L3 3M12 12L21.0001 3M12 12L3 21.0001"
                        id="Vector"
                        stroke="#000000"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </g>
                  </svg>
                </button>
              </td>
            </tr>
            <tr>
              <td>
                date
              </td>
              <td
                class="_created_542de3"
              >
                date
              </td>
              <td
                class="_ip_542de3"
              >
                10.0.0.2
              </td>
              <td>
                user_agent_full
              </td>
              <td>
                <button
                  class="_button_a7f9df _red_a7f9df  "
                  data-testid="revoke-session-btn"
                >
                  <svg
                    class="_sessionDeleteButton_542de3"
                    fill="none"
                    height="800px"
                    viewBox="0 0 24 24"
                    width="800px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      id="Menu / Close_LG"
                    >
                      <path
                        d="M21 21L12 12M12 12L3 3M12 12L21.0001 3M12 12L3 21.0001"
                        id="Vector"
                        stroke="#000000"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </g>
                  </svg>
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    class="_row_65046f"
  >
    <div
      class="_TokenList_6bec32"
    >
      <h2
        class="_headline_8da35a"
      >
        api_tokens
      </h2>
      <div
        class="_tableContainer_6bec32"
      >
        <table
          class="_table_6bec32"
        >
          <thead>
            <tr>
              <th>
                created
              </th>
              <th>
                name
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                date
              </td>
              <td>
                My Token 1
              </td>
              <td>
                <button
                  class="_button_a7f9df _red_a7f9df  "
                  data-testid="revoke-token-btn"
                >
                  <svg
                    class="_deleteButton_6bec32"
                    fill="none"
                    height="800px"
                    viewBox="0 0 24 24"
                    width="800px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      id="Menu / Close_LG"
                    >
                      <path
                        d="M21 21L12 12M12 12L3 3M12 12L21.0001 3M12 12L3 21.0001"
                        id="Vector"
                        stroke="#000000"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </g>
                  </svg>
                </button>
              </td>
            </tr>
            <tr>
              <td>
                date
              </td>
              <td>
                My Token 2
              </td>
              <td>
                <button
                  class="_button_a7f9df _red_a7f9df  "
                  data-testid="revoke-token-btn"
                >
                  <svg
                    class="_deleteButton_6bec32"
                    fill="none"
                    height="800px"
                    viewBox="0 0 24 24"
                    width="800px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      id="Menu / Close_LG"
                    >
                      <path
                        d="M21 21L12 12M12 12L3 3M12 12L21.0001 3M12 12L3 21.0001"
                        id="Vector"
                        stroke="#000000"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </g>
                  </svg>
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        <button
          class="_button_a7f9df   "
        >
          new_token
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
